### ISSUE TITLE
PSP ISO and CSO support?

### ISSUE THREAD

**Issue by Masamune3210** [GerbilSoft/rom-properties#213]
Would it be possible to add support for ISO and CSO files from the PSP? As far as I know, ISO files are just standard iso images while CSO are compressed image files. Maybe have boxart or the internal icon? Its located at PSP_GAME/ICON0.PNG along with the other pictures which represent the background when the game is selected on the menu and any alternate icons

**Comment by GerbilSoft** [comment]
I don't currently have any PSP disc images available for testing, so that'd be something I'd need to get.

rom-properties does have a basic ISO-9660 parser, and AFAIK PSP discs uses ISO-9660 and not UDF, so that should be fairly simple to implement.

CSO format might take a bit more effort, since that's zlib-compressed AFAIK.

For boxart, I'd need an external image database that allows access to the images using the game's internal ID.

**Comment by Masamune3210** [comment]
I can supply images if you need them of freshly ripped games straight from my collection, or you can source them yourself but iirc you are correct in that it's a standard iso. I know that you can rebuild a psp iso using standard tools so it can't be that different than stock iso. Boxart might be the hardest to track down but games have internal to them, icons, bgm for the game when you have it selected in the menu, background for the same operation, and in some cases differing save icons as well

**Comment by DankRank** [comment]
I looked at a bunch of ISOs I have and these are the relevant files:
```
/PSP_GAME/SYSDIR/UPDATE/PARAM.SFO - metadata for the firmware update
/PSP_GAME/ICON0.PNG \
/PSP_GAME/ICON1.PMF  \
/PSP_GAME/SND0.AT3    } - XMB Content Information Files (only ICON0.PNG is required)
/PSP_GAME/PIC0.PNG   /
/PSP_GAME/PIC1.PNG  /
/PSP_GAME/PARAM.SFO - metadata for the game
/UMD_DATA.BIN - contains disc id, some unknown stuff, and disc type
```
The ISO9660 System ID and Application were `PSP GAME` for all ISOs that I checked. (Probably shouldn't rely on that for detection though)

I also looked at a single UMD Video disc, it had its icon/metadata in `/UMD_VIDEO/` instead of `/PSP_GAME/`, and its System ID/Application was `UMD VIDEO`. The firmware update was still located at `/PSP_GAME/SYSDIR/UPDATE/`

Some info sources:
https://www.psdevwiki.com/ps3/Content_Information_Files#Games_PSP
https://www.psdevwiki.com/ps3/PARAM.SFO (the parameters there are PS3 specific, but the format is shared with PSP, as well as Vita and PS4)
https://www.psdevwiki.com/ps3/Eboot.PBP#PARAM.SFO (the parameters for PSP)

By the way, EBOOT.PBP is what's used for the digital downloads and homebrew on the PSP. It has the same sort of icon/metadata embedded in it.

**Comment by Masamune3210** [comment]
Would it be possible to parse PKG and pup files as well? PKG files are installable applications for PS3 and pup files are system updates for psp Vita PS3 and ps4

**Comment by DankRank** [comment]
Yes.

https://www.psdevwiki.com/ps3/PKG_files
https://github.com/euss/ps3tools/blob/master/ungpkg.c
(pkg contains ICON0.PNG, PARAM.SFO, etc. in the root. Packages that contain demo+full versions store metadata for the full version in `C00/`)

https://www.psdevwiki.com/ps3/Playstation_Update_Package_(PUP)
https://github.com/euss/ps3tools/blob/master/pupunpack.c
Test files: https://www.playstation.com/en-us/support/system-updates/
System updates for PSP are EBOOT.PBP, not PUP

**Comment by Masamune3210** [comment]
PS3 iso files are set up the same way iirc, Sony really hasn't changed anything file wise since the psp other than minor revisions so one bit of code could be possibly made to support most of it

**Comment by GerbilSoft** [comment]
Mostly implemented as of commit b9840512efcbf3f1520b03064049db00fed9bdd8.

`PARAM.SFO` isn't parsed, but the icon is thumbnailed and some other stuff is handled. CISO, ZISO, JISO, and DAX formats are supported, with zlib, lz4, and lzo compression where applicable.
